FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs

```diff
@@ -0,0 +1,117 @@
+        [HttpPost("like")]
+        public async Task<IActionResult> LikeNotificationAsync(string loggedInUser, string triggerUser, int postId)
+        {
+            return await CreateNotificationAsync(loggedInUser, triggerUser, postId, "like");
+        }
+
+        /// <summary>
+        /// API call from other services to create a follow notification. The user that was followed will get a notification.
+        /// </summary>
+        /// <param name="loggedInUser">email of user who has been followed (followee)</param>
+        /// <param name="triggerUser">email of the user who followed (follower)</param>
+        /// <param name="profileId">Profile ID of the trigger user</param>
+        /// <returns>Ok if accepted, BadRequest if fail</returns>
+        [HttpPost("follow")]
+        public async Task<IActionResult> FollowNotificationAsync(string loggedInUser, string triggerUser, int profileId)
+        {
+            return await CreateNotificationAsync(loggedInUser, triggerUser, profileId, "follow");
+        }
+
+        /// <summary>
+        /// Helper method to create the notification.
+        /// </summary>
+        /// <param name="loggedInUser">email of user that will receive the notification</param>
+        /// <param name="triggerUser">email of user that triggered the notification event</param>
+        /// <param name="linkId">Id of the resource to link to with the notification</param>
+        /// <param name="notificationType">The type of the notification (comment/follow/like)</param>
+        /// <returns></returns>
+        private async Task<IActionResult> CreateNotificationAsync(string loggedInUser, string triggerUser, int linkId, string notificationType)
+        {
+            Notification notification = new Notification
+            {
+                Type = new System.Collections.Generic.KeyValuePair<string, int>(notificationType, linkId),
+                LoggedInUserId = loggedInUser,
+                TriggerUserId = triggerUser,
+                Date = DateTime.Now
+            };
+
+            string methodName = "";
+            if (notificationType == "comment")
+            {
+                methodName = "CommentNotification";
+            }
+            else if (notificationType == "like")
+            {
+                methodName = "LikeNotification";
+            }
+            else if (notificationType == "follow")
+            {
+                methodName = "FollowNotification";
+            }
+
+            try
+            {
+                await notificationsRepo.CreateNotificationAsync(notification);
+                User user = await userRepo.GetUserAsync(loggedInUser);
+
+                foreach (var connection in user.Connections)
+                {
+                    await hub.Groups.AddToGroupAsync(connection, user.Email);
+                }
+
+                await hub.Clients.Group(user.Email).SendAsync(methodName, notification);
```

Comment:
i don't know signalr but are you sure this is right? you have to rebuild a group out of a user's connections every time just to send him something? and you have to do it slowly, one by one? sometimes it's unavoidable but you usually want to avoid `await` inside a loop at all costs

Reply:
The way it is currently set up, we have to do it this way. This sends the notification to all different places the same user is connected from (laptop, mobile, etc.).

Reply:
@ibnetahir1 at least add some comments giving some context that would help future development work to possibly improve this aspect. maybe reference links if you based your code/design on something you read online.